crates/cxx-qt-gen/src/generator/cpp/signal.rs

```diff
@@ -18,53 +22,61 @@ use syn::Result;
-struct SelfValue<'a> {
-    ident: &'a str,
-    ty: &'a str,
+    /// name with type of parameters
+    named_types: String,
+    /// name with type of parameters including self
+    named_types_with_self: String,
+    /// Raw types of the parameters including self
+    types_with_self: String,
+    /// Raw ::std::move values of the parameters including self
+    values_with_self: String,
+    values_with_self_as_this: String,
 }
 
 /// From given parameters, mappings, and self value constructor the combined parameter lines
 fn parameter_types_and_values(
     parameters: &[ParsedFunctionParameter],
     cxx_mappings: &ParsedCxxMappings,
-    self_value: SelfValue,
+    self_ty: &str,
 ) -> Result<Parameters> {
-    let mut parameter_types_closure = vec![];
-    let mut parameter_values_closure = vec![];
+    let mut parameter_named_types_with_self = vec![];
+    let mut parameter_types_with_self = vec![];
+    let mut parameter_values_with_self = vec![];
 
     for parameter in parameters {
         let cxx_ty = syn_type_to_cpp_type(&parameter.ty, cxx_mappings)?;
         let ident_str = parameter.ident.to_string();
-        parameter_types_closure.push(format!("{cxx_ty} {ident_str}",));
-        parameter_values_closure.push(format!("::std::move({ident_str})"));
+        parameter_named_types_with_self.push(format!("{cxx_ty} {ident_str}",));
+        parameter_types_with_self.push(cxx_ty.clone());
+        parameter_values_with_self.push(format!("::std::move({ident_str})"));
     }
 
-    let parameters_types_signal = parameter_types_closure.join(", ");
+    let parameter_named_types = parameter_named_types_with_self.join(", ");
+    let mut parameter_values_with_self_as_this = parameter_values_with_self.clone();
 
     // Insert the extra argument into the closure
-    parameter_types_closure.insert(0, format!("{ty}&", ty = self_value.ty));
-    parameter_values_closure.insert(0, self_value.ident.to_owned());
+    let self_ty = cxx_mappings.cxx(self_ty);
+    parameter_named_types_with_self.insert(0, format!("{self_ty}& self"));
+    parameter_types_with_self.insert(0, format!("{self_ty}&"));
+    parameter_values_with_self.insert(0, "self".to_owned());
+    parameter_values_with_self_as_this.insert(0, "*this".to_owned());
 
     Ok(Parameters {
-        types_closure: parameter_types_closure.join(", "),
-        types_signal: parameters_types_signal,
-        values_closure: parameter_values_closure.join(", "),
+        named_types: parameter_named_types,
+        named_types_with_self: parameter_named_types_with_self.join(", "),
+        types_with_self: parameter_types_with_self.join(", "),
+        values_with_self: parameter_values_with_self.join(", "),
+        values_with_self_as_this: parameter_values_with_self_as_this.join(", "),
     })
 }
 
 /// Generate C++ blocks for a free signal on an existing QObject (not generated by CXX-Qt), eg QPushButton::clicked
 pub fn generate_cpp_free_signal(
```

Comment:
reuse free generation for normal signal generation, eg make everything go via free functions